v2/apiserver/internal/core/jobs.go

```diff
@@ -190,6 +220,111 @@ func NewJobsService(
 	}
 }
 
+func (j *jobsService) Create(
+	ctx context.Context,
+	eventID string,
+	jobName string,
+	job Job,
+) error {
+	event, err := j.eventsStore.Get(ctx, eventID)
+	if err != nil {
+		return errors.Wrapf(err, "error retrieving event %q from store", eventID)
+	}
+	if _, ok := event.Worker.Jobs[jobName]; ok {
+		return &meta.ErrConflict{
+			Type: "Job",
+			ID:   jobName,
+			Reason: fmt.Sprintf(
+				"Event %q already has a job named %q.",
+				eventID,
+				jobName,
+			),
+		}
+	}
+
+	// Perform some validations...
+
+	// Determine if ANY of the job's containers:
+	//   1. Use shared workspace
+	//   2. Run in privileged mode
+	//   3. Mount the host's Docker socket
+	var useWorkspace = job.Spec.PrimaryContainer.UseWorkspace
+	var usePrivileged = job.Spec.PrimaryContainer.Privileged
+	var useDockerSocket = job.Spec.PrimaryContainer.UseHostDockerSocket
+	for _, sidecarContainer := range job.Spec.SidecarContainers {
+		if sidecarContainer.UseWorkspace {
+			useWorkspace = true
+		}
+		if sidecarContainer.Privileged {
+			usePrivileged = true
+		}
+		if sidecarContainer.UseHostDockerSocket {
+			useDockerSocket = true
+		}
+	}
+
+	// Fail quickly if any job is trying to run privileged or use the host's
+	// Docker socket, but isn't allowed to per worker configuration.
+	if usePrivileged &&
+		(event.Worker.Spec.JobPolicies == nil ||
+			!event.Worker.Spec.JobPolicies.AllowPrivileged) {
+		return &meta.ErrAuthorization{
+			Reason: "Worker configuration forbids jobs from utilizing privileged " +
+				"containers.",
+		}
+	}
+	if useDockerSocket &&
+		(event.Worker.Spec.JobPolicies == nil ||
+			!event.Worker.Spec.JobPolicies.AllowDockerSocketMount) {
+		return &meta.ErrAuthorization{
+			Reason: "Worker configuration forbids jobs from mounting the Docker " +
+				"socket.",
+		}
+	}
+
+	// Fail quickly if the job needs to use shared workspace, but the worker
+	// doesn't have any shared workspace.
+	if useWorkspace && !event.Worker.Spec.UseWorkspace {
+		return &meta.ErrConflict{
+			Reason: "The job requested access to the shared workspace, but Worker " +
+				"configuration has not enabled this feature.",
+		}
+	}
+
+	// Set the initial status
+	job.Status = &JobStatus{
+		Phase: JobPhasePending,
+	}
+
+	project, err := j.projectsStore.Get(ctx, event.ProjectID)
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"error retrieving project %q from store",
+			event.ProjectID,
+		)
+	}
+
+	if err = j.jobsStore.Create(ctx, eventID, jobName, job); err != nil {
+		return errors.Wrapf(
+			err, "error saving event %q job %q in store",
+			eventID,
+			eventID,
+		)
+	}
+
+	if err = j.substrate.ScheduleJob(ctx, project, event, jobName); err != nil {
+		return errors.Wrapf(
```

Comment:
There isn't... and that's an outstanding issue. Brigade coordinates between three other systems on the backend-- a database, a queue, and Kubernetes. There isn't a good way that I have found do distribute "transactions" across all three of those systems. i.e. No good way that I have discovered yet to make an API call an all-or-nothing where it all works or it all fails. I talked this over with @technosophos a few weeks ago and we agreed a good starting point was to do things in a way such that when something goes wrong, you're left in an "acceptable" state, where "acceptable" means you can figure out where things went wrong and there's a path of some kind to recovery (even if it means cloning the even to try the whole thing again). This is a big part of why I introduced the STARTING state earlier this week-- so that if anything ever gets stuck, more fine-grained states help us figure out where.

Reply:
Also @technosophos and I had agreed that performance and stability optimizations can come post-alpha, with the more immediate goal being to get the APIs defined and working well enough that work on gateways and on Kashti can proceed.

Reply:
Great, thanks for the background and context @krancour .  The reasoning sounds good to me!